Checking recursive APIs schemas result in StackOverflowException

Example: https://github.bus.zalan.do/setanta/api-reviews/pull/1/files

Stack is something like:
```
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_171]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_171]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_171]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_171]
        at de.zalando.zally.rule.RulesValidator.invoke(RulesValidator.kt:60) ~[classes!/:na]
        at de.zalando.zally.rule.RulesValidator.validate(RulesValidator.kt:40) ~[classes!/:na]
        at de.zalando.zally.rule.CompositeRulesValidator.validate(CompositeRulesValidator.kt:14) ~[classes!/:na]
        at de.zalando.zally.apireview.ApiViolationsController.validate(ApiViolationsController.kt:42) ~[classes!/:na]
...
Caused by: java.lang.StackOverflowError: null
        at io.swagger.v3.oas.models.media.StringSchema.hashCode(StringSchema.java:95) ~[swagger-models-2.0.3.jar!/:2.0.3]
        at java.util.Objects.hashCode(Objects.java:98) ~[na:1.8.0_171]
        at java.util.HashMap$Node.hashCode(HashMap.java:297) ~[na:1.8.0_171]
        at java.util.AbstractMap.hashCode(AbstractMap.java:530) ~[na:1.8.0_171]
        at java.util.Arrays.hashCode(Arrays.java:4146) ~[na:1.8.0_171]
        at java.util.Objects.hash(Objects.java:128) ~[na:1.8.0_171]
        at io.swagger.v3.oas.models.media.Schema.hashCode(Schema.java:795) ~[swagger-models-2.0.3.jar!/:2.0.3]
        at java.util.Arrays.hashCode(Arrays.java:4146) ~[na:1.8.0_171]
        at java.util.Objects.hash(Objects.java:128) ~[na:1.8.0_171]
        at io.swagger.v3.oas.models.media.ArraySchema.hashCode(ArraySchema.java:83) ~[swagger-models-2.0.3.jar!/:2.0.3]
```

--- a/server/src/main/java/de/zalando/zally/util/OpenApiUtil.kt
+++ b/server/src/main/java/de/zalando/zally/util/OpenApiUtil.kt
@@ -4,6 +4,7 @@ import io.swagger.v3.oas.models.OpenAPI
 import io.swagger.v3.oas.models.media.Schema
 import io.swagger.v3.oas.models.parameters.Parameter
 import io.swagger.v3.oas.models.responses.ApiResponse
+import java.util.Objects
 
 data class HeaderElement(
     val name: String,
@@ -62,9 +63,9 @@ fun OpenAPI.getAllSchemas(): Collection<Schema<Any>> = this.components?.schemas.
  * Traverses the schemas and returns all included schemas
  * @return a collection of all transitive schemas
  */
-fun OpenAPI.getAllTransitiveSchemas(): Set<Schema<Any>> {
+fun OpenAPI.getAllTransitiveSchemas(): Collection<Schema<Any>> {
     fun isPrimitive(schema: Schema<Any>): Boolean = schema.properties.orEmpty().isEmpty()
-    val collector = mutableSetOf<Schema<Any>>()
+    val collector = mutableMapOf<Int, Schema<Any>>()
 
     tailrec fun traverseSchemas(schemasToScan: Collection<Schema<Any>>) {
         if (schemasToScan.isEmpty()) return
@@ -72,13 +73,13 @@ fun OpenAPI.getAllTransitiveSchemas(): Set<Schema<Any>> {
         val primitiveSchemas = schemasToScan.filter { isPrimitive(it) }
         val nonPrimitiveSchemas = schemasToScan.filterNot { isPrimitive(it) }
 
-        collector.addAll(primitiveSchemas)
+        primitiveSchemas.forEach { schema -> collector[schema.customHash()] = schema }
         traverseSchemas(nonPrimitiveSchemas.flatMap { it.properties.values })
     }
 
     traverseSchemas(this.getAllSchemas())
 
-    return collector
+    return collector.values
 }
 
 /**
@@ -115,3 +116,13 @@ fun OpenAPI.getAllParameters(): Map<String, Parameter> = this.components?.parame
         it.readOperations()
             .flatMap { it.parameters.orEmpty().mapNotNull { it.name to it } }
     }
+
+/**
+ * Calculates custom hash to avoid calling the hash of the parent schema.
+ * E.g. io.swagger.v3.oas.models.media.ArraySchema#hashCode() calls super#hashCode()
+ * which fails for recursive data structures (endless loop -> StackOverflow)
+ */
+private fun Schema<Any>.customHash(): Int = Objects.hash(title, multipleOf, maximum, exclusiveMaximum, minimum,
+    exclusiveMinimum, maxLength, minLength, pattern, maxItems, minItems, uniqueItems, maxProperties, minProperties,
+    required, type, not, properties, additionalProperties, description, format, `$ref`, nullable, readOnly, writeOnly,
+    example, externalDocs, deprecated, xml, extensions, discriminator)

--- a/server/src/test/java/de/zalando/zally/util/OpenApiUtilTest.kt
+++ b/server/src/test/java/de/zalando/zally/util/OpenApiUtilTest.kt
@@ -15,6 +15,8 @@ import io.swagger.v3.oas.models.responses.ApiResponse
 import io.swagger.v3.oas.models.responses.ApiResponses
 import org.junit.Test
 import org.assertj.core.api.Assertions.assertThat
+import de.zalando.zally.getOpenApiContextFromContent
+import org.intellij.lang.annotations.Language
 
 class OpenApiUtilTest {
 
@@ -330,5 +332,40 @@ class OpenApiUtilTest {
 
         assertThat(schemas).isNotNull
         assertThat(schemas).isEmpty()
+    }
+
+    @Test
+    fun `getAllTransitiveSchemas should be able to cope with recursive schemas`() {
+        @Language("YAML")
+        val spec = """
+openapi: 3.0.1
+paths:
+  /products/{product_id}:
+    get:
+      responses:
+        200:
+          content:
+            application/json:
+              schema:
+                ${'$'}ref: "#/components/schemas/FacetGetResponse"
+components:
+  schemas:
+    FacetGetResponse:
+      allOf:
+        - ${'$'}ref: "#/components/schemas/ProductResource"
+    ProductResource:
+      allOf:
+        - type: object
+          properties:
+            children:
+              type: array
+              items:
+                ${'$'}ref: "#/components/schemas/ProductResource"
+        """.trimIndent()
+        val context = getOpenApiContextFromContent(spec)
+
+        val schemas = context.api.getAllTransitiveSchemas()
+
+        assertThat(schemas).hasSize(2)
     }
 }

